README.md
# README rSWeeP



The current release can be installed from Bioconductor:

```r
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("rSWeeP")
```

See the [website](https://camilapperico.github.io/) for additional information and tutorials.

### Citation

Please cite this package as follows:

De Pierri, Camilla Reginatto, Ricardo Voyceik, Letı́cia Graziela Costa Santos de Mattos, Mariane Gonçalves Kulik, Josué Oliveira Camargo, Aryel Marlus Repula de Oliveira, Bruno Thiago de Lima Nichio, et al. 2020. “SWeeP: Representing Large Biological Sequences Datasets in Compact Vectors.” Scientific Reports 10 (1): 1–10. (URL: https://doi.org/10.1038/s41598-019-55627-4). 

### License

This package is released in the public domain under the creative commons license [CC0](https://tldrlegal.com/license/creative-commons-cc0-1.0-universal). 


NEWS
Changes in version 0.99.0 (2019-05-15)
+ Submitted to Bioconductor

Changes in version 0.99.4 (2020-01-21)
+ The package no longer relies on seqinr, now operates only with Biostrings functions and variables
+ The  Title and the Description have been updated to be more precise in describing the package.
+ Despcritional comments have been added to all codes.
+ The sWeeP functions now accept  AAStringset as well file names as a valid entry
+ The sWeeP function now uses an S4 Generic and S4 methods
+ The user input variable baseMatrix has now a  validating step on the code
+ The sWeeP function now uses Biostrings::readAAStringSet for FASTA input.
+ All functions now only use message(...) to communicate with users.
+ All 1:length(...) instances have been replaced with seq_len
+ On all codes, the function tidy_source() from the formatR package has been used to standardizing code formating.
+ All operations that do not affect have to the package working been removed.
+ Some functions and variables naming has been changed aiming to follow the lowerCamelCase convention (the sWeeP function still as it's, due it is the method initials)
+ aa2int() has been replaced by a vector inside  the aa2num2()
+ The data set datastring has been removed.
+ fas2mat receives now an AAStringSet as entry
+ Instead of reading the R object datastring and writing it as a FASTA file, a FASTA file has been included directly in inst/extdata/
+ A bigger FASTA has been used in the "Quick Start" section
+ All non-R package source files have been removed

Changes in version 0.99.5 (2020-01-21)
+ An even bigger FASTA has been used in the "Quick Start" section

Changes in version 0.99.6 (2020-01-21)
+ Error removed

Changes in version 0.99.7 (2020-01-21)
+ Warning removed

Changes in version 0.99.12 (2020-02-08)
+ 2000+ sequences multifasta error removed

Changes in version 1.3.2 (2020-11-24)
+ baseMatrix error fix

R/SWeeP.R
#' @title A vectorial comparative method to amino acid sequence.
#' @name sWeeP
#' @description The "Spaced Words Projection (SWeeP)" is a method for
#' representing biological sequences using compact vectors. SWeeP uses the
#' spacedwords concept by scanning sequences and generating indices
#' to create a higherdimensional matrix of occurrences that is later projected into a
#' smaller randomly oriented orthonormal base (PIERRI, 2019). This way the resulting
#' matrix  will conserve the comparational data but will have a selectable size
#' @param xfas A AAStringSet or a FASTA format file
#' @param baseMatrix A orthonormal matrix with 160.000 coordinates
#' @details The SWeeP method was developed to favor the comparison between
#' complete proteomic sequences and to assist in machine learning analyzes. This
#' method is based on the concept of spaced words, which are used to scan
#' biological sequences and project them into  matrix of occurrences, favoring the manipulation
#' of the data. The sWeeP function can project a matrix n by m, where n is the number of
#' sequences in the analized \code{xfas} and m is the number of columns
#' in \code{baseMatrix} matrix.
#' @return A matrix resulted by the projection of the sequences in \code{xfas}
#' in the \code{baseMatrix} matrix
#' @author Danrley R. Fernandes.
#' @references Pierri,C. R. et al. SWeeP: Representing large biological
#' sequences data sets in compact vectors. Scientific Reports, accepted in
#' December 2019.doi: 10.1038/s41598-019-55627-4.
#' @examples
#' baseMatrix <- orthBase(160000,10)
#' path <- system.file(package = "rSWeeP", "extdata", "exdna.fas")
#' return <- sWeeP(path,baseMatrix)
#' distancia <- dist(return, method = "euclidean")
#' tree <- hclust(distancia, method="ward.D")
#' plot(tree, hang = -1, cex = 1)
#' @rdname sWeeP
#' @export

##assuring that exposed functions of your package interoperate with standard Bioconductor objects by an S4 Generic and S4 methods
setGeneric(
  "sWeeP",
  function(xfas, baseMatrix) standardGeneric("sWeeP"),
  signature = "xfas"
)
#' @rdname sWeeP
setMethod("sWeeP", "character", function(xfas, baseMatrix) {
  aa = Biostrings::readAAStringSet(xfas,"fasta", use.names = FALSE)
  sWeeP(aa, baseMatrix)
})
#' @rdname sWeeP
setMethod("sWeeP", "AAStringSet", function(xfas, baseMatrix) {
  ##validating user input.
  stopifnot(methods::is(baseMatrix, "matrix"))
  ##if the data sequence has more than than 2000 specimens it's divide to process in pieces of 2000 at time
  if (length(xfas) > 2001) {
    ##define the number of fragmentation needed
    barra <- round(length(xfas)/2000)
    blocksToConvert <- length(xfas)/barra
    block <- round(length(xfas)/round(length(xfas)/blocksToConvert))
    ##create the matrix who will contain the projection
    Bigprojected <- matrix(0,0,dim(baseMatrix)[2])
      ##position defines the start of projection a at position 1
     position <- 1
      ##runs the projection thru all the matrix
    for (i in  1:barra) {
      message("running ", i, " of ", barra)
      ##verifi if the next matrix area isn't outside limits
      if (position + block < length(xfas)) {
        projected <- fas2mat(xfas[position:(position + block)]) %*% baseMatrix
        Bigprojected<-rbind(Bigprojected,projected)
        position <- position + block + 1
      }
      ##if the next matrix area is outside limits performs the projection only until end of the matrix
      else {
        projected <- fas2mat(xfas[position:length(xfas)]) %*% baseMatrix
        Bigprojected<-rbind(Bigprojected,projected)
      }
      projected <- Bigprojected
    }
  }
  ##if the data sequence has less than than 2000 specimens it's run the projection directly
  else {
    matrizid <- fas2mat(xfas)
    projected <- matrizid %*% baseMatrix
  }
  ##returns the projection matrix
  projected
})
